### PR TITLE
Directly show "Forgot password?" link, fix #6808

### LINKED ISSUE #6808: Password Reset is not user friendly
The way to reset your password is not user friendly. I just added 20 users to my instance, and had the people reset their passwords. Suffice it to say this did not go well.

One of the biggest problems is this "button"

password reset

It is not understood as a button or link. People see this message, but they do not think they could click it.
-> **It needs to be distinguishable as something that can be done**

Also, people who _know_ they forgot their password don't ever reach this message. Normal people don't entertain the idea of _intentionally_ giving a bad password, just to _possibly_ reset their passwords.
-> **Nextcloud needs an explicit "forgot your password?" link**

### PR DESCRIPTION
**Pull request by jancborchardt**
We still need some JS so that when you click »Forgot password?«:

- [x] The password field is hidden
- [x] The log in button text is changed to »Reset password«
- [x] After the password is reset, the view is morphed back

Ref #6808 cc @ccoenen @nextcloud/javascript

### CODE REVIEW

**Comment by ccoenen** [comment]
Why would this involve JavaScript at all? Ideally this would be a seperate url that I could point users to. As in "oh, I see, please go to *this URL* , enter your user name and you'll receive an email".

**Comment by jancborchardt** [comment]
@ccoenen that is not mutually exclusive. :) The point is it’s a nicer experience for people from the log in page if it’s an animation rather than a hard page refresh. It can have a different URL though, like with a hash.

File: core/css/guest.css [hunk 1]
```diff
@@ -384,11 +384,21 @@ form .warning input[type='checkbox']+label {
...
 	margin: 18px 5px 0 16px !important;
 }
-.remember-login-container {
+.remember-login-container,
+.lost-password-container {
 	display: inline-block;
 	margin: 10px 0;
 	text-align: center;
 	width: 100%;
+	text-shadow: 0 0 2px rgba(0, 0, 0, .4); // better readability on bright background
```
Comment (ccoenen) [inline core/css/guest.css]: this only works accidentally, `//` is not a valid comment in regular CSS. (found once more below at line 572).

**Comment by ccoenen** [comment]
when would this change in its current form have effect? I can see the `$_['canResetPassword']` in login.php, but as an outsider, I have no Idea what it actually means. I do not understand how/when the lost password container would be inserted.

**Comment by jancborchardt** [comment]
Pushed some JS code to hide and show the elements for the password reset. Can someone with more technical knowledge check it please? @skjnldsv @blizzz @MorrisJobke @juliushaertl :)

File: core/templates/login.php [hunk 2]
```diff
@@ -51,7 +51,7 @@
 			<label for="password" class="infield"><?php p($l->t('Password')); ?></label>
 		</p>
 
-		<?php if (!empty($_['invalidpassword']) && !empty($_['canResetPassword'])) { ?>
+		<?php if (!empty($_['invalidpassword']) && !empty($_['canResetPlost-passwordassword'])) { ?>
```
Comment (ccoenen) [inline core/templates/login.php]: this looks like a copy&paste error

Comment (jancborchardt) [reply core/templates/login.php]: Thanks, good catch.

**Comment by jancborchardt** [comment]
This is ready to review now @nextcloud/javascript @rullzer @MorrisJobke @blizzz please check this out :)

**Review by skjnldsv — DISMISSED** [review]
- When clicking the forgot password link then entering the username:
   - the submit button stays disabled.
   - using the enter key result by an error in the console: `An invalid form control with name='password' is not focusable.`
- When entering the username THEN clicking the forgot password link:
   - We could use a loader
   - The form is directly sent, I think we should probably wait for a manual action from the user

**Comment by ccoenen** [comment]
I can't tell from the code alone (and I can't include this patch in my instance right now), can somebody tell me, if this contains a mechanism to _directly_ start which the lost password view? (Without the user needing to click or something)

Something, where I can send users to _directly_ reset their passwords. A URL i can stuff into an email easily. Something like these:

- https://www.paypal.com/authflow/password-recovery/
- https://de.wordpress.com/wp-login.php?action=lostpassword
- https://www.facebook.com/login/identify?ctx=recover&lwv=110
- https://fyp.ebay.de/EnterUserInfo

(you get the idea)

**Comment by skjnldsv** [comment]
@ccoenen no, there is a few clicks needed.

**Comment by ccoenen** [comment]
Thank you, in that case, I would like to request that this is added to this pull request as well. This is something I mentioned in the original ticket #6808, and that I think is rather important.

**Comment by skjnldsv** [comment]
@ccoenen Well, it does directly show the "forgot password" button now, so it's just one click. You don't have to try to login to display the message now at least :)

**Comment by jancborchardt** [comment]
As said, I would need a JS person to take it from here then. :)

**Comment by skjnldsv** [comment]
Okay, I'm still working on it, will finish when home from work:
- [ ] Message error css
- [ ] Loader fix and addition for lostpassword
- [ ] Toggle fix
- [ ] Fix the dual lostpassword buttons after a failed login
- [ ] The page reload on click after a failed login
- [ ] Fix disabled state of submit
- [ ] Fix enter key of form for lostpassword

**Comment by skjnldsv** [comment]
Everything is fixed! :)
@nextcloud/designers @jancborchardt 

Honestly, we could really use a restructuration of the page, it looks a but messy. But it works and looks great ;)

Please check the following situations:
- Reset password with unknown user
- Reset with existing username
- Fail login then reset password with existing user
- Fail login then reset password with unknown user

Remember, every failed attempt will trigger a bruteforce protection, to test, remember to clean your `oc_bruteforce_attempts` table.

**Review by skjnldsv — APPROVED** [review]

**Comment by jancborchardt** [comment]
Thanks a lot @skjnldsv! :) Can we please have a review @nextcloud/javascript? :)

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
acceptance tests for login failed, but maybe it was a hiccup - I restated them (failed ones: https://drone.nextcloud.com/nextcloud/server/2642/239)

**Comment by skjnldsv** [comment]
Nope, failure is related to this pr! :)

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
I get a "CSRF check failed" exception. Seems that you messed around with the hidden secret that is in the form to avoid such attacks. Please have a look

**Comment by skjnldsv** [comment]
Acceptance test passed, I restarted for an unrelated failure.

**Comment by skjnldsv** [comment]
Few bugs on android
- [ ] too long button text
- [ ] Grey background when scrolled up

screenshot_20171108-114305

**Comment by MorrisJobke** [comment]
> I get a "CSRF check failed" exception. Seems that you messed around with the hidden secret that is in the form to avoid such attacks. Please have a look

@skjnldsv I have seen this after entering the new password

**Comment by jancborchardt** [comment]
Should we merge for 13 and fix the issues separately? Cause they are mobile only.

**Comment by MorrisJobke** [comment]
> Should we merge for 13 and fix the issues separately? Cause they are mobile only.

I'm fine with that.

**Comment by jancborchardt** [comment]
@skjnldsv a fix for the first point you mentioned is at https://github.com/nextcloud/server/pull/7167. Second one I can’t reproduce, can you open an issue?

**Comment by skjnldsv** [comment]
@jancborchardt Okay! :)

**Comment by ccoenen** [comment]
thanks to everyone involved! Looking forward to the next release :-)

**Pull request merged**